### PR TITLE
Implements conditionals resolving service

### PR DESCRIPTION
**Pull request by custode**
Depends on #852

### CODE REVIEW

File: app/services/course/conditional/conditional_resolving_service.rb [hunk 1]
```diff
@@ -0,0 +1,48 @@
+class Course::Conditional::ConditionalResolvingService
+  class << self
+    # Resolve the conditionals for the given course user
+    #
+    # @param[CourseUser] course_user The course user with conditionals to be resolved
```
Comment (lowjoel) [inline app/services/course/conditional/conditional_resolving_service.rb]: space before [

File: app/models/course/condition.rb [hunk 2]
```diff
@@ -14,6 +14,16 @@ class Course::Condition < ActiveRecord::Base
   ].freeze
 
   class << self
+    # Finds all the conditionals for the given course.
+    #
+    # @param [Course] course
```
Comment (lowjoel) [inline app/models/course/condition.rb]: Want to document the course attribute?

File: app/models/course/condition.rb [hunk 3]
```diff
@@ -14,6 +14,16 @@ class Course::Condition < ActiveRecord::Base
...
+    # @return [Object] acts_as_conditionals objects belonging to the given course
```
Comment (lowjoel) [inline app/models/course/condition.rb]: You might want to specify a common duck-typed method that all acts_as_conditional objects implement.

File: app/services/course/conditional/conditional_resolving_service.rb [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+    delegate :resolve, to: :new
+  end
+
+  # Resolve the conditionals for the given course user
+  #
+  # @param[CourseUser] course_user The course user with conditionals to be resolved
```
Comment (lowjoel) [inline app/services/course/conditional/conditional_resolving_service.rb]: space before [ also

File: app/services/course/conditional/conditional_resolving_service.rb [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+  def resolve(course_user)
+    # Retrieve the conditional dependency graph for the course user
+    conditional_graph = conditional_graph_for(course_user.course)
```
Comment (lowjoel) [inline app/services/course/conditional/conditional_resolving_service.rb]: Some of these variables, like course_user and course can be instance variables, which are assigned in an initialize. The Service object only is accessible from one thread, and it is encapsulated, so that is fine. Saves you passing objects around as arguments

File: spec/services/course/conditional/conditional_resolving_service_spec.rb [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+        it 'adds the resolved achievement conditional to the course user' do
+          allow(course_user).to receive(:level_number).and_return(2)
+          # TODO: Check that the achievement was added to the course user once conditional's API
+          # are added in
+          expect(subject).to contain_exactly(achievement)
+        end
+      end
+
+      context 'when course user do not satisfy the level condition' do
+        # TODO: Check that course user do not have the achievement once conditional's APi are
```
Comment (lowjoel) [inline spec/services/course/conditional/conditional_resolving_service_spec.rb]: API

File: app/models/course/condition.rb (hunk 3 above)
Comment (custode) [inline app/models/course/condition.rb]: Sounds good. Will do it in a separate PR to fix all of them.
